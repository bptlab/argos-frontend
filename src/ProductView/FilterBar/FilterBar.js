import React, { Component } from 'react';
import Filter from './Filter/Filter.js';

class FilterBar extends Component {
    render() {
        return (
            <div className="filter-bar container">
                <div className="form-group row">
                    {this.props.filter.map((filter, index) =>
                        <Filter key={filter.id} id={index} value={filter.value}
                                onChangeFilterInput={this.props.onChangeFilterInput}/>
                    )}
                </div>
            </div>
        );
    }
}

export default FilterBar;
