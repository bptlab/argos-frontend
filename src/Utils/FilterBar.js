import React, {Component} from "react";
import config from "./../config/config";
import {css} from "aphrodite";
import SearchBar from "./SearchBar";
import {Row, Col} from "react-grid-system";

/**
 * @fileOverview
 * Usage:
 *     <FilterBar
 *         _Required_
 *             onFiltersChange      {callback}  -> Called if any filter in bar changed
 *         _Optional_
 *             autoCompleteSource   {array}     -> Passed down, please refer {@link SearchBar}
 *             useColumns           {bool}      -> Passed down, please refer {@link SearchBar}
 *             styles
 *     />
 */
class FilterBar extends Component {
	constructor () {
		super();
		this.styles = {
			primaryBorderColor: {
				borderColor: config.colors.primary,
			}
		};

		this.state = {
			filter: [{id: '0', value: '', column: null}],
			lastFilterId: 0,
		};

		this.onFilterChange = this.onFilterChange.bind(this);
	}

	onFilterChange(filter) {
		let updatedFilters = this.state.filter;
		const currentFilterIndex = this.getPositionOfFilterInState(filter);

		updatedFilters[currentFilterIndex] = filter;

		if (!filter.value && !filter.column) {
			updatedFilters.splice(currentFilterIndex, 1);
		}

		updatedFilters = this.assureThatEmptyFilterExists(updatedFilters);

		this.setState({
				filter: updatedFilters
			},
			() => this.props.onFiltersChange(this.state.filter)
		);
	}

	getPositionOfFilterInState(filter) {
		const filterIds = this.state.filter.map(oldFilter => {
			return oldFilter.id;
		});
		return filterIds.indexOf(filter.id);
	}

	assureThatEmptyFilterExists(filters) {
		const emptyFilterExists = filters.some(filter => {
			return !filter.value && !filter.column;
		});

		if(!emptyFilterExists) {
			const newFilterId = this.state.lastFilterId + 1;
			const newFilter = {id: `${newFilterId}`, value: '', column: null};
			filters = filters.concat([newFilter]);
			this.setState({
				lastFilterId: newFilterId,
			});
		}

		return filters;
	}

	render() {
		return (
			<Row className={css(this.props.styles)}>
				{this.state.filter.map(filter => {
					return(
						<Col sm={4} key={filter.id}>
							<SearchBar
								id={filter.id}
								value={filter.value}
								column={filter.column}
								autoCompleteSource={this.props.autoCompleteSource}
								onInputChange={this.onFilterChange}
							    useColumns={this.props.useColumns}
							/>
						</Col>
					);
				})}
			</Row>
		);
	}
}

export default FilterBar;
