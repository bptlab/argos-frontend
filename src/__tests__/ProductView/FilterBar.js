import React from 'react';
import FilterBar from '../../ProductView/FilterBar/FilterBar.js';
import renderer from 'react-test-renderer';
import ReactTestUtils from 'react-addons-test-utils'

let instance, callbackMock, filter1, filter2;

beforeAll(() => {
    callbackMock = jest.fn();
    filter1 = {id: '0', value: 'foo', column: null};
    filter2 = {id: '1', value: '', column: null};
});

test("Rendering of FilterBar", () => {
    const component = renderer.create(
        <FilterBar
            ref={(child) => {instance = child}}
            onInputChange={callbackMock}
            filter={[filter1]} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Remove button is displayed correctly", () => {
    renderer.create(
        <FilterBar
            ref={(child) => {instance = child}}
            onInputChange={callbackMock}
            filter={[filter1,filter2]} />
    );

    expect(instance.decideOnRemoveButton(filter1)).toBeTruthy();
    expect(instance.decideOnRemoveButton(filter2)).toBeFalsy();
});

test("Handle click of remove button", () => {
    const rendered = ReactTestUtils.renderIntoDocument(<FilterBar
        ref={(child) => {instance = child}}
        onInputChange={callbackMock}
        filter={[filter1,filter2]} />);
    const deleteButton = ReactTestUtils.findRenderedDOMComponentWithTag(rendered, "button");
    ReactTestUtils.Simulate.click(deleteButton);

    expect(callbackMock).toBeCalled();
});

