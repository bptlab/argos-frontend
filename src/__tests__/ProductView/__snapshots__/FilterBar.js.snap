exports[`test Rendering of FilterBar 1`] = `
<div
  className="filter-bar container">
  <div
    className="form-group row">
    <div
      className="input-field input-group col-xs-12 col-sm-4 col-lg-3">
      <span
        className="input-group-addon"
        id="search-addon">
        <i
          className="fa fa-search" />
      </span>
      <input
        className="form-control"
        onChange={[Function]}
        placeholder="Search..."
        type="text"
        value={undefined} />
    </div>
  </div>
</div>
`;
